sycl/source/detail/program_impl.cpp

```diff
@@ -42,7 +43,9 @@ program_impl::program_impl(
     DevicesSorted = sort_devices_by_cl_device_id(MDevices);
   }
   check_device_feature_support<info::device::is_linker_available>(MDevices);
+  vector_class<std::unique_lock<std::mutex>> Locks;
   for (const auto &Prg : ProgramList) {
+    Locks.emplace_back(Prg->MMutex);
```

Comment:
Deadlock is possible here. Its probability is rather small though. Reproducing use-case is pretty unreal: link program out of multiple programs in one thread whilst these programs are compiled/linked in other threads.

Smth like [std::lock](https://en.cppreference.com/w/cpp/thread/lock) may be used. Though it is templated on the sequence of mutexes. In general it employs some backoff.

Reply:
Good catch. We also run into a problem here if the user tries to link two copies of the same program. How about locking the mutexes in a sorted manner and checking for program duplication?

Reply:
I suggest employing atomic variable for state of the program. A single (per-process) pair of `std::condition_variable` and `std::mutex` will be required also.

Though sorting programs may be helpful where applicable (i.e. where the order of programs provided by the user is not required to be preserved).